### PR TITLE
Add support for serialization of large caches

### LINKED ISSUE #45: Serializing Large LRUs hangs/StackOverflows
The following code hangs on my machine and performs a stackoverflow on another:
```julia
using LRUCache
using Serialization

# Create an LRU cache with a capacity of 100k elements
cache = LRU{Int, Int}(maxsize=100_000)

# Populate the cache with dummy data
for i in 1:100_000
    cache[i] = i
end

serialize("cache.jls", cache)
```
I think this is because when there are too many linked nodes, the stack overflows before all the nodes can be serialized. I'm going to be writing some custom serialization logic to handle large LRUs, was wondering if upstream might want the changes.

### PR DESCRIPTION
**Pull request by jarbus**
Fixes #45, where serialization of large LRU might result in a stack overflow. Does not automatically load Serialization, but uses extensions to overwrite Serialization.{serialize,deserialize} if Serialization is loaded, available in julia 1.9+.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,13 +1,20 @@
 name = "LRUCache"
 uuid = "8ac3fa9e-de4c-5943-b1dc-09c6b5f20637"
-version = "1.6.0"
+version = "1.6.1"
 
 [compat]
-julia = "1"
+julia = "≥ 1.9"
```
Comment (ericphanson) [inline Project.toml]: you could use [this strategy](https://pkgdocs.julialang.org/v1/creating-packages/#Transition-from-normal-dependency-to-extension) to maintain support for earlier Julia versions. That would make Serialization a direct dep on pre-1.9 and a weakdep on 1.9+

File: test/serializationtests.jl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    # Check that the cache is the same
+    @test cache.maxsize == deserialized_cache.maxsize
+    @test cache.currentsize == deserialized_cache.currentsize
+    @test cache.hits == deserialized_cache.hits
+    @test cache.misses == deserialized_cache.misses
+    @test cache.by == deserialized_cache.by
+    @test cache.finalizer == deserialized_cache.finalizer
+    @test cache.keyset.length == deserialized_cache.keyset.length
+    @test length(cache.dict) == length(deserialized_cache.dict)
```
Comment (ericphanson) [inline test/serializationtests.jl]: This would test all the same pairs are present in the cache without needing internals:
```suggestion
    @test issetequal(collect(cache), collect(deserialized_cache))
```
(though probably good to check the internal details too, as you do)

File: Project.toml (hunk 1 above)
Comment (jarbus) [reply Project.toml]: Brilliant, didn't know about this. Thanks!

**Comment by jarbus** [comment]
Thanks for the review @ericphanson, committed the changes you suggested!

File: ext/SerializationExt.jl [hunk 3]
```diff
@@ -0,0 +1,83 @@
+module SerializationExt
+export serialize, deserialize
```
Comment (ericphanson) [inline ext/SerializationExt.jl]: I don't think this does anything (can't have exports from an extension, IIUC):
```suggestion
```

File: ext/SerializationExt.jl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+using LRUCache
+using Serialization
+
+# Serialization of large LRUs causes a stack overflow error, so we 
+# create a custom serializer that represents LinkedNodes as Ints
+function Serialization.serialize(s::AbstractSerializer, lru::LRU{K, V}) where {K, V}
+    # Create a mapping from memory address to id
+    node_map = Dict{Ptr, Int}()
```
Comment (ericphanson) [inline ext/SerializationExt.jl]: Rather than using Ptr's, I think it is simpler (and maybe more robust) to use an `IdDict`:
```suggestion
    # Create a mapping from object to id. Here we use `IdDict` to use object identity as the hash.
    node_map = IdDict{LinkedNode{K}, Int}()
```
then this can be indexed just as `node_map[node] = id` rather than using `pointer_from_objref`.

File: test/serializationtests.jl [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    cache = LRU{Int, Int}(maxsize=100_000)
+
+    # Populate the cache with dummy data
+    for i in 1:1_000_000
+        cache[i] = i+1
+    end
+    serialize("cache.jls", cache)
+    deserialized_cache = deserialize("cache.jls")
+    rm("cache.jls")
```
Comment (ericphanson) [inline test/serializationtests.jl]: We can avoid the filesystem by using an IOBuffer:
```suggestion
    io = IOBuffer()
    serialize(io, cache)
    seekstart(io)
    deserialized_cache = deserialize(io)
```

**Review by ericphanson — COMMENTED** [review]
left a few more comments; overall this looks pretty good to me, though I've never written a serialize method before so not sure if there's things to look out for here

File: ext/SerializationExt.jl (hunk 3 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: Interesting, I needed these before to get some code I was using to access the proper serialization method, but for some reason that's not happening anymore.

File: ext/SerializationExt.jl (hunk 4 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: This is great, thanks

File: test/serializationtests.jl (hunk 5 above)
Comment (jarbus) [reply test/serializationtests.jl]: Also a good change, thanks

**Review by ericphanson — APPROVED** [review]
lgtm, but it would be great if someone more familiar with serialization could review as well

File: ext/SerializationExt.jl [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    dict = Dict{K, Tuple{V, LRUCache.LinkedNode{K}, Int}}()
+    # Create the nodes, but don't link them yet
+    for (key, (value, id, s)) in dict_with_ids
+        nodes[id] = LRUCache.LinkedNode{K}(key)
+        dict[key] = (value, nodes[id], s)
+    end
+    # Link the nodes
+    for (idx, node) in enumerate(nodes)
+        node.next = nodes[idx % n_nodes + 1]
+        node.prev = nodes[idx == 1 ? n_nodes : idx - 1]
```
Comment (Jutho) [inline ext/SerializationExt.jl]: I think you can write these two right hand sides using `mod1(idx+1, n_nodes)` and `mod1(idx-1, n_nodes)`, which I would find more clear.

File: test/serializationtests.jl [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    @test cache.misses == deserialized_cache.misses
+    @test cache.by == deserialized_cache.by
+    @test cache.finalizer == deserialized_cache.finalizer
+    @test cache.keyset.length == deserialized_cache.keyset.length
+    @test issetequal(collect(cache), collect(deserialized_cache))
+    # Check that the cache has the same keyset
+    c_node = cache.keyset.first
+    d_node = deserialized_cache.keyset.first
+    for i in 1:cache.keyset.length
+        c_node.val == d_node.val || @test false
```
Comment (Jutho) [inline test/serializationtests.jl]: Why not simply `@test c_node.val == d_node.val` ?

File: test/serializationtests.jl [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        c_node = c_node.next
+        d_node = d_node.next
+    end
+    # Check that the cache keys, values, and sizes are the same
+    for (key, (c_value, c_node, c_s)) in cache.dict
+        d_value, d_node, d_s = deserialized_cache.dict[key]
+        c_value == d_value || @test false
+        c_node.val == d_node.val || @test false
+        c_s == d_s || @test false
```
Comment (Jutho) [inline test/serializationtests.jl]: Same question for these tests.

**Comment by Jutho** [comment]
This looks great, thanks. I've also added some minor comments and questions.

File: test/serializationtests.jl (hunk 7 above)
Comment (jarbus) [reply test/serializationtests.jl]: This is just personal taste, for a large dict of 100k entries, I don't want to add 100k tests that compare each element. I just want to have one test that fails if any element is different, which I believe these do

File: ext/SerializationExt.jl (hunk 6 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: Didn't know about the `mod1` function, thanks! That's much cleaner.

**Comment by jarbus** [comment]
All great feedback so far, I've learned an unexpected amount about Julia from this simple PR. Thanks!

File: test/serializationtests.jl (hunk 7 above)
Comment (Jutho) [reply test/serializationtests.jl]: Sure, that is true. On the other hand, the test could also have just been with a handful of elements in the LRU cache I believe. I haven't timed it and believe this is all very fast because it is just `Int`s, but the 100000 did jump to the eye as some large number to have as a simple test case.

Comment (ericphanson) [reply test/serializationtests.jl]: for small caches, the regular serialization method works fine and doesn't stackoverflow, so we need a big one to really test the new method

Comment (jarbus) [reply test/serializationtests.jl]: ^ exactly right

Comment (Jutho) [reply test/serializationtests.jl]: Well, yes and no, since the new method is now always called, irrespective of the size of the cache. And it is written in such a way that it should not suffer from the same flaws, so I would think that if it passes the test for a smaller cache, that provides sufficient guarantees. But I am fine with the current tests.

Is it clear what was causing the default serialisation strategy to fail? My guess of why it was entering an infinite loop would apply irrespective of the size of the cache, so that is not consistent with the original method working for small caches.

**Comment by Jutho** [comment]
Do you want another review? I can ask a person in our group that recently implemented some serialisation things for some other cache structures that we have in our code, but he is only available from Friday onwards. I am also fine with having this merged as is.

**Comment by jarbus** [comment]
I'm fine with either, doesn't hurt to get another set of eyes on it and I've already pointed all my code to my branch anyways

File: ext/SerializationExt.jl [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    end
+    @assert id == length(lru) == lru.keyset.length == length(lru.dict)
+    # By this point, the first node has id 1 and the last node has id length(lru)
+    # so when deserializing, we can infer the order by the id
+    # Create the dict with ids instead of nodes
+    dict = Dict{K, Tuple{V, Int, Int}}()
+    for (key, (value, node, s)) in lru.dict
+        id = node_map[node]
+        dict[key] = (value, id, s)
+    end
```
Comment (lkdvos) [inline ext/SerializationExt.jl]: I would maybe change `s` here to something else, for example `sz`. `s` is also used for the `AbstractSerializer`, and while the for loop introduces a new scope, this might cause unnecessary confusion.
```suggestion
    for (key, (value, node, sz)) in lru.dict
        id = node_map[node]
        dict[key] = (value, id, sz)
    end
```

File: ext/SerializationExt.jl [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    first_node = node = lru.keyset.first
+    node_map[node] = id
+    # Go through the rest of the nodes in the cycle and create a mapping
+    node = node.next
+    while node != first_node
+        id += 1
+        node_map[node] = id
+        node = node.next
+    end
+    @assert id == length(lru) == lru.keyset.length == length(lru.dict)
```
Comment (lkdvos) [inline ext/SerializationExt.jl]: In principle this can be condensed by using the iterator interface that is defined for `CyclicOrderedSet`. It should produce the exact same code, but reads a little easier.
```suggestion
    for (id, node) in enumerate(lru.keyset)
        node_map[node] = id
    end
```

File: ext/SerializationExt.jl [hunk 11]
```diff
@@ -0,0 +1,82 @@
+module SerializationExt
+using LRUCache
+using Serialization
+
+# Serialization of large LRUs causes a stack overflow error, so we 
+# create a custom serializer that represents LinkedNodes as Ints
+function Serialization.serialize(s::AbstractSerializer, lru::LRU{K, V}) where {K, V}
+    # Create a mapping from memory address to id
+    node_map = IdDict{LRUCache.LinkedNode{K}, Int}()
```
Comment (lkdvos) [inline ext/SerializationExt.jl]: I think you can add a size hint here:
```suggestion
    node_map = IdDict{LRUCache.LinkedNode{K}, Int}()
    sizehint!(node_map, length(lru))
```

File: test/serializationtests.jl (hunk 7 above)
Comment (lkdvos) [reply test/serializationtests.jl]: Maybe as a middle ground, the following gives a single (two) test(s), but still tests all values are equal:
```julia
@test length(cache.keyset) == length(deserialized_cache.keyset)
@test all(((c_val, d_val),) -> c_val == d_val, zip(cache.keyset, deserialized_cache.keyset))
```

**Review by lkdvos — COMMENTED** [review]
I think this looks very nice! I left some minor comments which only really change readability, and thus might be a bit subjective as well. Feel free to discard those if you would disagree.

As a more general comment, I think it is possible to circumvent the exchange of the dictionary with linked nodes for a new dictionary, which is keeping an integer for the order. AFAIK, serialization of a dictionary happens by just serializing `(key, value)` pairs sequentially. In that sense, they already have an order, which you could use as the id.

Conveniently, iterating over an `LRU` actually gives you the `key => value` pairs in the exact order you want, because of how iteration of the keyset is happening. Thus, you could consider something along the lines of the following, and avoid the intermediate additional dictionary:
[Serialization link](https://github.com/JuliaLang/julia/blob/3ed49fdb3a9c500d0134320151614d39d517def3/stdlib/Serialization/src/Serialization.jl#L367)
```julia
write(s.io, Int32(length(lru)))
for (k, (val, node, sz)) in lru
    serialize(s, k)
    serialize(s, val)
    serialize(s, sz)
end
```

For the deserialization, it should be possible to then just deserialize, and similar to how you implemented it wrap the keys back in a `LinkedNode`, linking them along the way, because you can now be sure that you are traversing them in order. (special casing the first and last node). (Note that this snippet does not include special cases for dicts that are empty or have a single element)
[link for deserialization](https://github.com/JuliaLang/julia/blob/3ed49fdb3a9c500d0134320151614d39d517def3/stdlib/Serialization/src/Serialization.jl#L1602)
```julia
dict = Dict{K, Tuple{V, LRUCache.LinkedNode{K}, Int}}()
n = read(s.io, Int32)
sizehint!(dict, n)
# first entry
k = deserialize(s)
first = node = LRUCache.LinkedNode{K}(k)
val = deserialize(s)
sz = deserialize(s)
dict[k] = (val, node, sz)
# middle entries
for i in 2:n-1
    prev = node
    k = deserialize(s)
    node = LRUCache.LinkedNode{K}(k)
    prev.next = node
    node.prev = prev
    val = deserialize(s)
    sz = deserialize(s)
    dict[k] = (val, node, sz)
end
# last node
prev = node
k = deserialize(s)
node = LRUCache.LinkedNode{K}(k)
prev.next = node
node.prev = prev

val = deserialize(s)
sz = deserialize(s)
dict[k] = (val, node, sz)

node.next = first
first.prev = node
```

I think I am still missing some things to make the keyset etc, and I did not test anything yet. I think this should result in a smaller file size, and maybe a bit better performance. Of course, depending on the use-case, this might all be overkill, and your implementation is definitely also viable!

File: ext/SerializationExt.jl (hunk 10 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: Thanks! Great suggestion.

File: ext/SerializationExt.jl (hunk 11 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: Brilliant! Never heard about `sizehint!` before.

**Comment by lkdvos** [comment]
Woops, seems like I was too fast in suggesting the while loop to for loop change, the iterator indeed only returns the values, not the nodes themselves. My apologies!

**Comment by jarbus** [comment]
Thanks for the in-depth feedback @lkdvos! I'll look into implementing your suggested refactor. Also I suppose I need to undo that last commit now as well haha

**Comment by jarbus** [comment]
@lkdvos Finished working on your recommended changes, serialization code is much cleaner now. Thanks for your input!

**Comment by Jutho** [comment]
Looks good to me; maybe @lkdvos can take a final look and then we can merge.

**Review by ericphanson — COMMENTED** [review]
I think it would be worth testing roundtripping an entirely empty `lru`, as well as one with one entry. Since those look like edge cases in the deserialization implementation. Maybe:
```julia
lru = LRU{Int,Int}(; maxsize=10)
for n in 0:5
        if n > 0
             lru[n] = n
        end
        io = IOBuffer()
        serialize(io, lru)
        seekstart(io)
        lru_rt = deserialize(io)
        @test lru isa LRU{Int,Int}
        @test issetequal(collect(lru), collect(lru_rt))
end
```

One other edge case I can think of is mutable values with shared object identity. Like
```julia
lru = LRU(; maxsize=5)
a = b = [1]
lru[1] = a
lru[2] = b
@test lru[1] === lru[2]
# now roundtrip it and check that the above `===` still holds
```

**Comment by jarbus** [comment]
Good point, honestly not sure how to handle the multiple references to mutable values case, though.

**Comment by ericphanson** [comment]
The regular serializer handles it, so I was hoping it would just work

**Comment by jarbus** [comment]
Great feedback @ericphanson, worked as expected. I was able to clean everything up, code is much nicer now.

File: ext/SerializationExt.jl [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    sizehint!(dict, currentsize)
+    # Create node chain
+    first = nothing
+    node = nothing
+    for i in 1:currentsize
+        prev = node
+        k = deserialize(s)
+        node = LRUCache.LinkedNode{K}(k)
+        val = deserialize(s)
+        sz = by(val)::Int
```
Comment (lkdvos) [inline ext/SerializationExt.jl]: I think this is not entirely safe to just recompute the size. In principle, as the `LRU`s are mutable, it is allowed to change the `by` field halfway through filling up the cache. This would then no longer correctly infer the sizes. It's an easy fix though, you just need to (de)serialize (key, val, sz) together.

File: test/serializationtests.jl [hunk 13]
```diff
@@ -0,0 +1,49 @@
+using Serialization
+@testset "Large Serialize and Deserialize" begin
+
+    cache = LRU{Int, Int}(maxsize=100_000)
+
+    # Populate the cache with dummy data
+    num_entries_to_test = [0, 1, 2, 3, 4, 5, 100_000, 1_000_000]
+    for i in 0:maximum(num_entries_to_test)
+        i ∈ num_entries_to_test || continue
```
Comment (ericphanson) [inline test/serializationtests.jl]: this looks a bit strange to me, should the `cache[i] = i+1` go before the `continue` line? So that we're still filling the cache even when we aren't testing it

Comment (jarbus) [reply test/serializationtests.jl]: Good catch, not sure how I missed this.

File: ext/SerializationExt.jl (hunk 12 above)
Comment (jarbus) [reply ext/SerializationExt.jl]: I did this because the size isn't returned when we iterate through the LRU, and it was suggested to iterate through the LRU for serialization. You are right though, good point - I'll work on another method to iterate through the LRU

**Review by lkdvos — APPROVED** [review]

**Comment by jarbus** [comment]
@Jutho @ericphanson are there any other changes you recommend before merging?

**Comment by ericphanson** [comment]
Sorry, forgot about this. I think it’s good to go!
